File: sdk/memorygraphsdk/exceptions.py
"""
Exceptions for MemoryGraph SDK.
"""


class MemoryGraphError(Exception):
    """Base exception for MemoryGraph SDK errors."""

    def __init__(self, message: str, status_code: int | None = None):
        super().__init__(message)
        self.message = message
        self.status_code = status_code


class AuthenticationError(MemoryGraphError):
    """Raised when API key is invalid or missing."""

    def __init__(self, message: str = "Invalid or missing API key"):
        super().__init__(message, status_code=401)


class RateLimitError(MemoryGraphError):
    """Raised when rate limit is exceeded."""

    def __init__(self, message: str = "Rate limit exceeded"):
        super().__init__(message, status_code=429)


class NotFoundError(MemoryGraphError):
    """Raised when a resource is not found."""

    def __init__(self, message: str = "Resource not found"):
        super().__init__(message, status_code=404)


class ValidationError(MemoryGraphError):
    """Raised when request validation fails."""

    def __init__(self, message: str = "Validation error"):
        super().__init__(message, status_code=400)


class ServerError(MemoryGraphError):
    """Raised when server returns an error."""

    def __init__(self, message: str = "Server error"):
        super().__init__(message, status_code=500)

File: ts/src/sdk/exceptions.ts
/**
 * Exceptions for the MemoryGraph SDK.
 *
 * Mirrors the Python SDK exception hierarchy. Each error optionally carries
 * the HTTP status code that triggered it.
 */

export class MemoryGraphError extends Error {
  statusCode?: number;

  constructor(message: string, statusCode?: number) {
    super(message);
    this.name = "MemoryGraphError";
    if (statusCode !== undefined) {
      this.statusCode = statusCode;
    }
  }
}

export class AuthenticationError extends MemoryGraphError {
  constructor(message = "Invalid or missing API key") {
    super(message, 401);
    this.name = "AuthenticationError";
  }
}

export class RateLimitError extends MemoryGraphError {
  constructor(message = "Rate limit exceeded") {
    super(message, 429);
    this.name = "RateLimitError";
  }
}

export class NotFoundError extends MemoryGraphError {
  constructor(message = "Resource not found") {
    super(message, 404);
    this.name = "NotFoundError";
  }
}

export class ValidationError extends MemoryGraphError {
  constructor(message = "Validation error") {
    super(message, 400);
    this.name = "ValidationError";
  }
}

export class ServerError extends MemoryGraphError {
  constructor(message = "Server error") {
    super(message, 500);
    this.name = "ServerError";
  }
}
